substrate-archive-backend/src/read_only_backend/blockchain_backend.rs

```diff
@@ -33,22 +34,59 @@ use sp_runtime::{
 use crate::{
 	database::ReadOnlyDb,
 	read_only_backend::ReadOnlyBackend,
-	util::{self, columns},
+	util::{self, columns, read_db},
 };
 
 type ChainResult<T> = Result<T, BlockchainError>;
 
+#[derive(Debug, Encode, Decode)]
+struct ExtrinsicHeader {
+	hash: sp_core::H256,
+	data: Vec<u8>,
+}
+
 impl<Block: BlockT, D: ReadOnlyDb> BlockchainBackend<Block> for ReadOnlyBackend<Block, D> {
 	fn body(&self, id: BlockId<Block>) -> ChainResult<Option<Vec<<Block as BlockT>::Extrinsic>>> {
-		let res = util::read_db::<Block, D>(&*self.db, columns::KEY_LOOKUP, columns::BODY, id)
-			.map_err(|e| BlockchainError::Backend(e.to_string()))?;
-
-		match res {
-			Some(body) => match Decode::decode(&mut &body[..]) {
+		let body = match read_db(&*self.db, columns::KEY_LOOKUP, columns::BODY, id)
+			.map_err(|e| BlockchainError::Backend(e.to_string()))?
+		{
+			Some(body) => body,
+			None => return Ok(None),
+		};
+		match self.storage_mode {
+			TransactionStorageMode::BlockBody => match Decode::decode(&mut &body[..]) {
 				Ok(body) => Ok(Some(body)),
-				Err(_) => Err(BlockchainError::Backend("Could not decode extrinsics".into())),
+				Err(err) => return Err(BlockchainError::Backend(format!("Error decoding body: {}", err))),
+			},
+			TransactionStorageMode::StorageChain => match Vec::<ExtrinsicHeader>::decode(&mut &body[..]) {
+				Ok(index) => {
+					let extrinsics: ChainResult<Vec<Block::Extrinsic>> = index
+						.into_iter()
+						.map(|ExtrinsicHeader { hash, data }| {
+							let decode_result = if hash != Default::default() {
+								match self.db.get(columns::TRANSACTION, hash.as_ref()) {
+									Some(t) => {
+										let input = [&data[..], &t[..]].concat();
```

Comment:
Substrate has that funny `JoinInput` thing, I guess you went with this to cut down on dependencies? (Well, it's also not `pub` I guess).

I didn't read that code carefully, but you're confident concatenating is all we need here?

Reply:
Can confirm, added a test just to be sure it does the same thing